tests/spec_decode/e2e/test_eagle_correctness.py

```diff
@@ -0,0 +1,268 @@
+"""This docstring details important information on the testing methodology.
+
+Most of the tests rely on "greedy equality", where we expect the output of
+speculative decoding on a sequence to exactly match the output of normal non-
+speculative decoding.
+
+Since speculative decoding with rejection sampling guarantees that the output
+distribution matches the target model's output distribution (up to hardware
+numerics, see https://arxiv.org/pdf/2302.01318.pdf), we can expect greedy
+equality.
+
+However, we still need to verify below scenario could be passed:
+    * Batch size 1 greedy equality
+    * Batch size >1 greedy equality
+    * Test greedy equality under preemption
+    * Test greedy equality under various number of speculative tokens.
+
+With those tests, we can say at least, EAGLE would not break the
+correctess for the target model outputs.
+"""
+
+import pytest
+
+from .conftest import run_greedy_equality_correctness_test
+
+# main model
+MAIN_MODEL = "JackFram/llama-68m"
+
+# speculative model
+SPEC_MODEL = "abhigoyal/vllm-eagle-llama-68m-random"
+
+# max. number of speculative tokens: this corresponds to
+# num_heads in the config.json of the speculator model.
+MAX_SPEC_TOKENS = 4
+
+# precision
+PRECISION = "float32"
```

Comment:
Is this the recommended configuration for production deployments? We may be hiding a correctness issue that appears in fp16.

Reply:
For prod, we can use FP16 as well and it should work fine for all practical purposes. However, I've noticed that when using FP16, the distribution of the target model differs (though minimally). This sometimes results in different tokens being selected by the target model depending on it being used as a verifier or an independent generator.

For a more concrete example, please take a look at this: https://github.com/vllm-project/vllm/pull/4978#discussion_r1627603023